src/ValueParsers/TimeParser.php

```diff
@@ -43,83 +44,113 @@ class TimeParser extends StringValueParser {
+			. '(?:T(\d{2}):(\d{2})(?::(\d{2}))?)?'                        //hour, minute and second
+			. 'Z?'                                                        //time zone
+			. '\s*\(?\s*' . CalendarModelParser::MODEL_PATTERN . '\s*\)?' //calendar model
+			. '\s*$@iu';                                                  //trailing spaces
+
+		if ( !preg_match( $pattern, $value, $matches ) ) {
+			throw new ParseException( 'Malformed time', $value, self::FORMAT_NAME );
+		} elseif ( $matches[2] < 60 && $matches[5] === '' ) {
+			throw new ParseException( 'Not enough information to decide if the format is YY-MM-DD',
+				$value, self::FORMAT_NAME );
+		} elseif ( $matches[3] > 12 ) {
+			throw new ParseException( 'Month out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[4] > 31 ) {
+			throw new ParseException( 'Day out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[5] > 23 ) {
+			throw new ParseException( 'Hour out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[6] > 59 ) {
+			throw new ParseException( 'Minute out of range', $value, self::FORMAT_NAME );
+		} elseif ( $matches[7] > 62 ) {
+			throw new ParseException( 'Second out of range', $value, self::FORMAT_NAME );
+		}
+
+
+		$matches = array_slice( $matches, 1 );
+		$matches[0] = str_replace( "\xE2\x88\x92", '-', $matches[0] );
+
+		return $matches;
 	}
 
 	/**
-	 * @param array $timeParts with the following keys.
-	 *            sign, year, month, day, hour, minute, second, calendar
+	 * @param string[] $timeParts Array with index 0 => sign, 1 => year, 2 => month, etc.
 	 *
-	 * @return int precision as a TimeValue PRECISION_ constant
+	 * @return int Precision as a TimeValue::PRECISION_ constant
 	 */
-	private function getPrecisionFromTimeParts( $timeParts ) {
-		if( $timeParts['second'] !== '00' ) {
-			return TimeValue::PRECISION_SECOND;
-		}
-		if( $timeParts['minute'] !== '00' ) {
-			return TimeValue::PRECISION_MINUTE;
-		}
-		if( $timeParts['hour'] !== '00' ) {
-			return TimeValue::PRECISION_HOUR;
-		}
-		if( $timeParts['day'] !== '00' ) {
-			return TimeValue::PRECISION_DAY;
+	private function getPrecision( array $timeParts ) {
+		if ( intval( $timeParts[6] ) > 0 ) {
+			$precision = TimeValue::PRECISION_SECOND;
+		} elseif ( intval( $timeParts[5] ) > 0 ) {
+			$precision = TimeValue::PRECISION_MINUTE;
+		} elseif ( intval( $timeParts[4] ) > 0 ) {
+			$precision = TimeValue::PRECISION_HOUR;
+		} elseif ( intval( $timeParts[3] ) > 0 ) {
+			$precision = TimeValue::PRECISION_DAY;
+		} elseif ( intval( $timeParts[2] ) > 0 ) {
+			$precision = TimeValue::PRECISION_MONTH;
+		} else {
+			$precision = $this->getPrecisionFromYear( $timeParts[1] );
 		}
-		if( $timeParts['month'] !== '00' ) {
-			return TimeValue::PRECISION_MONTH;
+
+		$option = $this->getOption( self::OPT_PRECISION );
+
+		// It's impossible to increase precision via option, e.g. to month if no month is given
+		if ( is_int( $option ) && $option <= $precision ) {
```

Comment:
That's a special case, not touched and not relevant to this patch. Please open a ticket.